From 8b70e7f63c49fb73d18539550ebd067c001c2e1d Mon Sep 17 00:00:00 2001
From: Vincent Bernat <vincent@bernat.im>
Date: Thu, 12 Apr 2018 11:54:58 +0200
Subject: [PATCH] Disable IBRS in VMX, even if VM request it

A VM can play with SPEC_CTRL MSR even when the guest CPUID doesn't
have support for it. We don't want that. And in fact, we don't want
people to use this MSR at all as it is a performance sink. Nowaday,
other mitigations are available. Only bare-metal may need IBRS when
calling for firmwares.
---
 arch/x86/kvm/vmx.c  | 3 ++-
 include/linux/smp.h | 1 +
 kernel/sysctl.c     | 9 +++++++++
 3 files changed, 12 insertions(+), 1 deletion(-)

Index: pkg-kernel-4.4/arch/x86/kvm/vmx.c
===================================================================
--- pkg-kernel-4.4.orig/arch/x86/kvm/vmx.c
+++ pkg-kernel-4.4/arch/x86/kvm/vmx.c
@@ -2940,7 +2940,8 @@ static int vmx_set_msr(struct kvm_vcpu *
 		if (data & ~(SPEC_CTRL_IBRS | SPEC_CTRL_STIBP | SPEC_CTRL_SSBD))
 			return 1;
 
-		vcpu->arch.spec_ctrl = data;
+		if (!sysctl_ibrs_vmx_forbidden)
+			vcpu->arch.spec_ctrl = data;
 		break;
 	case MSR_IA32_CR_PAT:
 		if (vmcs_config.vmentry_ctrl & VM_ENTRY_LOAD_IA32_PAT) {
Index: pkg-kernel-4.4/include/linux/smp.h
===================================================================
--- pkg-kernel-4.4.orig/include/linux/smp.h
+++ pkg-kernel-4.4/include/linux/smp.h
@@ -59,6 +59,7 @@ int smp_call_function_single_async(int c
 /* indicate usage of IBRS to control execution speculation */
 extern int use_ibrs;
 extern u32 sysctl_ibrs_enabled;
+extern u32 sysctl_ibrs_vmx_forbidden;
 extern struct mutex spec_ctrl_mutex;
 
 static inline int __check_ibrs_inuse(void)
Index: pkg-kernel-4.4/kernel/sysctl.c
===================================================================
--- pkg-kernel-4.4.orig/kernel/sysctl.c
+++ pkg-kernel-4.4/kernel/sysctl.c
@@ -248,6 +248,8 @@ u32 sysctl_ibrs_enabled = 0;
 EXPORT_SYMBOL(sysctl_ibrs_enabled);
 u32 sysctl_ibpb_enabled = 0;
 EXPORT_SYMBOL(sysctl_ibpb_enabled);
+u32 sysctl_ibrs_vmx_forbidden = 0;
+EXPORT_SYMBOL(sysctl_ibrs_vmx_forbidden);
 
 /* The default sysctl tables: */
 
@@ -1244,6 +1246,13 @@ static struct ctl_table kern_table[] = {
 		.extra2         = &two,
 	},
 	{
+		.procname       = "ibrs_vmx_forbidden",
+		.data           = &sysctl_ibrs_vmx_forbidden,
+		.maxlen         = sizeof(unsigned int),
+		.mode           = 0644,
+		.proc_handler	= proc_dointvec,
+	},
+	{
 		.procname       = "ibpb_enabled",
 		.data           = &sysctl_ibpb_enabled,
 		.maxlen         = sizeof(unsigned int),
